clients/src/main/java/org/apache/kafka/common/requests/DescribeAclsResponse.java

```diff
@@ -17,164 +17,82 @@
+        this.data = data;
     }
 
-    public DescribeAclsResponse(Struct struct) {
-        this.throttleTimeMs = struct.get(THROTTLE_TIME_MS);
-        this.error = new ApiError(struct);
-        this.acls = new ArrayList<>();
-        for (Object resourceStructObj : struct.getArray(RESOURCES_KEY_NAME)) {
-            Struct resourceStruct = (Struct) resourceStructObj;
-            ResourcePattern pattern = RequestUtils.resourcePatternromStructFields(resourceStruct);
-            for (Object aclDataStructObj : resourceStruct.getArray(ACLS_KEY_NAME)) {
-                Struct aclDataStruct = (Struct) aclDataStructObj;
-                AccessControlEntry entry = RequestUtils.aceFromStructFields(aclDataStruct);
-                this.acls.add(new AclBinding(pattern, entry));
-            }
-        }
+    public DescribeAclsResponse(Struct struct, short version) {
+        this.data = new DescribeAclsResponseData(struct, version);
     }
 
     @Override
     protected Struct toStruct(short version) {
         validate(version);
-
-        Struct struct = new Struct(ApiKeys.DESCRIBE_ACLS.responseSchema(version));
-        struct.set(THROTTLE_TIME_MS, throttleTimeMs);
-        error.write(struct);
-
-        Map<ResourcePattern, List<AccessControlEntry>> resourceToData = new HashMap<>();
-        for (AclBinding acl : acls) {
-            resourceToData
-                .computeIfAbsent(acl.pattern(), k -> new ArrayList<>())
-                .add(acl.entry());
-        }
-
-        List<Struct> resourceStructs = new ArrayList<>();
-        for (Map.Entry<ResourcePattern, List<AccessControlEntry>> tuple : resourceToData.entrySet()) {
-            ResourcePattern resource = tuple.getKey();
-            Struct resourceStruct = struct.instance(RESOURCES_KEY_NAME);
-            RequestUtils.resourcePatternSetStructFields(resource, resourceStruct);
-            List<Struct> dataStructs = new ArrayList<>();
-            for (AccessControlEntry entry : tuple.getValue()) {
-                Struct dataStruct = resourceStruct.instance(ACLS_KEY_NAME);
-                RequestUtils.aceSetStructFields(entry, dataStruct);
-                dataStructs.add(dataStruct);
-            }
-            resourceStruct.set(ACLS_KEY_NAME, dataStructs.toArray());
-            resourceStructs.add(resourceStruct);
-        }
-        struct.set(RESOURCES_KEY_NAME, resourceStructs.toArray());
-        return struct;
+        return data.toStruct(version);
     }
 
     @Override
     public int throttleTimeMs() {
-        return throttleTimeMs;
+        return data.throttleTimeMs();
     }
 
     public ApiError error() {
-        return error;
+        return new ApiError(Errors.forCode(data.errorCode()), data.errorMessage());
     }
 
     @Override
     public Map<Errors, Integer> errorCounts() {
-        return errorCounts(error.error());
+        return errorCounts(Errors.forCode(data.errorCode()));
     }
 
     public Collection<AclBinding> acls() {
```

Comment:
We should probably remove this method. It makes it too easy to write highly inefficient code. See https://github.com/apache/kafka/pull/7725 for an alternative approach.

Reply:
Obsolutely! I've updated the logic to avoid that.